### PR TITLE
Introduce ability to minimize round-trips in CCS

### PR DESCRIPTION
**Pull request by javanna**
With #37566 we have introduced the ability to merge multiple search responses into one. That makes it possible to expose a new way of executing cross-cluster search requests, that makes CCS much faster whenever there is network latency between the CCS coordinating node and the remote clusters. The coordinating node can now send a single search request to each remote cluster, which gets reduced by each one of them. `from` + `size` results are requested to each cluster, and the reduce phase in each cluster is non final (meaning that buckets are not pruned and pipeline aggs are not executed). The CCS coordinating node performs an additional, final reduction, which produces one search response out of the multiple responses received from the different clusters.

This new execution path will be activated by default for any CCS request unless a scroll is provided or inner hits are requested as part of field collapsing. The search API accepts now a new parameter called `ccs_minimize_roundtrips` that allows to opt-out of the default behaviour.

Relates to #32125

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java [hunk 1]
```diff
@@ -190,35 +191,90 @@ protected void doExecute(Task task, SearchRequest searchRequest, ActionListener<
...
+                    final int size;
+                    final int trackTotalHitsUpTo;
+                    if (source == null) {
+                        from = SearchService.DEFAULT_FROM;
+                        size = SearchService.DEFAULT_SIZE;
+                        trackTotalHitsUpTo = SearchContext.DEFAULT_TRACK_TOTAL_HITS_UP_TO;
+                    } else {
+                        from = source.from() == -1 ? SearchService.DEFAULT_FROM : source.from();
+                        size = source.size() == -1 ? SearchService.DEFAULT_SIZE : source.size();
+                        trackTotalHitsUpTo = source.trackTotalHitsUpTo();
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java]: `source.trackTotalHitsUpTo()` is nullable in https://github.com/elastic/elasticsearch/pull/37466 so you'll need to check here too and replace with `SearchContext.DEFAULT_TRACK_TOTAL_HITS_UP_TO` if it's the case.

**Review by jimczi — COMMENTED** [review]
LGTM, I left a minor comment regarding another pr that will conflict with this one.

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java (hunk 1 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java]: good one, I will update once the linked PR is in.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java]: I have updated the PR accordingly.

**Review by jimczi — APPROVED** [review]

File: docs/reference/modules/cross-cluster-search.asciidoc [hunk 2]
```diff
@@ -273,3 +279,45 @@ GET /cluster_one:twitter,cluster_two:twitter,twitter/_search <1>
...
+the search executes as if all shards were part of the same cluster. The
+coordinating node sends one request to each shard involved, each shard
+executes the query and returns its own results which are then reduced (and
+fetched, depending on the <<search-request-search-type, search type>>) by the
+coordinating node. The `local` reduce mode is beneficial whenever there is very
+ low network latency between the coordinating node and the remote clusters
+ involved, as it treats all shards the same, at the cost of sending many
+ requests to each remote cluster.
+
+- `auto`: `remote` is used whenever possible. In case a scroll is provided,
```
Comment (s1monw) [inline docs/reference/modules/cross-cluster-search.asciidoc]: so what happens if I specify `remote` and I have inner hits?

Comment (s1monw) [reply docs/reference/modules/cross-cluster-search.asciidoc]: I wonder if we have to have a `mode` at all. can't we simply go with `local_reduce: true|false` if it's true we force local if not we try remote if possible?

**Review by s1monw — CHANGES_REQUESTED** [review]
I left a question

File: docs/reference/modules/cross-cluster-search.asciidoc (hunk 2 above)
Comment (javanna) [reply docs/reference/modules/cross-cluster-search.asciidoc]: > so what happens if I specify remote and I have inner hits?

If you specify remote and you provide a scroll or request inner hits as part of fields collapsing, you get back an error. I think it's better to let users know that what they have requested is not supported when they are explicit in their request. 

> I wonder if we have to have a mode at all. can't we simply go with local_reduce: true|false if it's true we force local if not we try remote if possible?

I think that having remote (or false) act only as a preference is confusing, as users are requesting something that may not be possible. Ideally, if somebody sets `local_reduce: false` (or `reduce_mode: remote`), we would never go and fan out to the shards. It sounds weird to me to have an option act as a proper on/off behaviour and the other one as a preference, unless they have different names.

If we want to have only two values and get rid of the `auto` option, we could have `reduce_mode: always_local | prefer_remote` which I find  much more self-explanatory than a boolean. Yet I am not sure it is much different from what we have now in the PR.

Comment (javanna) [reply docs/reference/modules/cross-cluster-search.asciidoc]: or we could also remote `auto` to `prefer_remote`.

Comment (javanna) [reply docs/reference/modules/cross-cluster-search.asciidoc]: we discussed this a bit with @jimczi and the team, we could go with a `boolean` but I would like to have some `force` part in the name of the option. We could go with `ccs_force_local_reduce` which defaults to false, or maybe even `ccs_optimize_for_latency` which defaults to true? I think I prefer the latter as it tells when such option should be used. Any other thoughts/preference?

**Comment by clintongormley** [comment]
> We could go with ccs_force_local_reduce which defaults to false, or maybe even ccs_optimize_for_latency which defaults to true? I think I prefer the latter as it tells when such option should be used.

With `ccs_optimize_for_latency` it is not clear to me which option I'd be opting for - i'd have to really understand the feature.  Also, given that we want to use remote reduce as the default because we think it'll be better, then i think `ccs_force_local_reduce: true` works as a good way to opt out of that behaviour.

**Comment by s1monw** [comment]
> With ccs_optimize_for_latency it is not clear to me which option I'd be opting for - i'd have to really understand the feature. Also, given that we want to use remote reduce as the default because we think it'll be better, then i think ccs_force_local_reduce: true works as a good way to opt out of that behaviour.

I am not sure if `ccs_force_local_reduce` explains what's going on. the reduce is not relevant here IMO. This is more about the network calls we do. Something like `css_minimize_roundtrips: true|false` would be more descriptive IMO.

**Comment by zuketo** [comment]
Does the term WAN help here? E.g. `ccs_wan_optimized: true|false`

"We're operating across a large network, let's keep this set to true" vs "We're operating within a local network, let's investigate changing this setting"

**Comment by javanna** [comment]
> I am not sure if ccs_force_local_reduce explains what's going on. the reduce is not relevant here IMO. This is more about the network calls we do. Something like css_minimize_roundtrips: true|false would be more descriptive IMO.

I agree with this, I like `ccs_minimize_roundtrips`, I think it explains the trade-off well, at the same time staying high-level and without mentioning reduction. I briefly chatted to @clintongormley and he also agrees. I will adapt the PR accordingly.

**Comment by s1monw** [comment]
thanks @javanna

**Comment by javanna** [comment]
heya @s1monw can you have another look? it should be ready.

**Comment by javanna** [comment]
run elasticsearch-ci/2

**Review by s1monw — APPROVED** [review]
LGTM thanks @javanna

**Pull request merged**